### DISCUSSION TITLE
Can't have cumulative rollover budget amounts

### DISCUSSION THREAD

**Discussion by ashemsay** [firefly-iii/firefly-iii#7289] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I can't have cumulative amounts for my rollover budget.
I set up a rollover budget on 2023-02-16 with a monthly amount of 150 and an initial ammount of 210. I spent 120.23€ the first month (2023-02) and 154.52€ the second month (2023-03).

I can go to the first month and see the correct amounts:
![firefly-budget-bug-february](https://user-images.githubusercontent.com/8015125/227997665-2bbb0099-d712-4a1d-8a07-118fc8dbd6b8.png)

The issue is on the second month, the leftover from the first month has not been carried out:
![firefly-budget-bug-march](https://user-images.githubusercontent.com/8015125/227998047-b207f159-8996-43f7-945d-521b72001ebc.png)

### Debug information

Debug information generated at 2023-03-27 00:00:00 Europe/Paris for Firefly III version **6.0.5**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.5 |
| Firefly III API | 2.0.1 |
| Build | 424, 18-03-2023 14:53:12 CET |
| DB version | 19 (exp. 19) |
| PHP | `8.2.3` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | Europe/Paris |
| Browser TZ | Europe/Paris |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | fr_FR |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | mysql |
| Trusted proxies (.env) | `**` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2023-02-16 00:00:00 |
| Session end | 2023-03-27 00:00:00 |
| Session first | 2023-02-16 00:00:00 |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/111.0 |

### Expected behaviour

I should see 150+89.77=239.77 budgeted for march.

### Additional info
I feel like there's a similarity with [this issue](https://github.com/firefly-iii/firefly-iii/issues/4290), but that one complains about the wrong left calculation, here I have an issue with the budgeted amount not being correctly set.

Note that I had not set up the cron job initially, but successfully ran and scheduled the execution through curl earlier today:

```json
{"recurring_transactions":{"job_fired":true,"job_succeeded":true,"job_errored":false,"message":"Recurring transactions cron job fired successfully."},"auto_budgets":{"job_fired":true,"job_succeeded":true,"job_errored":false,"message":"Auto-budget cron job fired successfully."}}
```

I could reproduce the issue on the demo site with the "car" budget by making sure there is some amount left at the end of february, it didn't get carried over to march.

**Comment by JC5** [comment]
I'm sorry, I can't replicate this issue. I did everything you did, and I end up with the correct amount for March 2023:

<img width="1392" alt="Screenshot 2023-03-27 at 20 44 40" src="https://user-images.githubusercontent.com/5889984/228036966-ae98ac3e-dd50-4a38-b903-bb7303caefde.png">

In my logs, I see the correct behavior and math:

```
[2023-03-27 20:41:11] local.INFO: Today (2023-03-01) is a magic day for monthly auto-budget #4 (part of budget #6 "TestBudget")  
[2023-03-27 20:41:11] local.DEBUG: Going to find a budget limit for budget #6 ("TestBudget") between 2023-03-01 and 2023-03-31  
[2023-03-27 20:41:11] local.DEBUG: Will now manage rollover for auto budget #4  
[2023-03-27 20:41:11] local.DEBUG: Current period is 2023-03-01-2023-03-31, so previous period is 2023-02-01-2023-02-28  
[2023-03-27 20:41:11] local.DEBUG: Going to find a budget limit for budget #6 ("TestBudget") between 2023-02-01 and 2023-02-28  
[2023-03-27 20:41:11] local.DEBUG: Budget limit exists for previous period.  
[2023-03-27 20:41:11] local.DEBUG: Now in FireflyIII\Repositories\Budget\OperationsRepository::sumExpenses  
[2023-03-27 20:41:11] local.DEBUG: Currency is "Euro".  
[2023-03-27 20:41:11] local.DEBUG: Found 0 journals with currency EUR.  
[2023-03-27 20:41:11] local.DEBUG: Spent in previous budget period (2023-02-01-2023-02-28) is -120.230000000000  
[2023-03-27 20:41:11] local.DEBUG: Total amount left for previous budget period is 89.770000000000  
[2023-03-27 20:41:11] local.INFO: The amount left is positive, so the new amount will be 239.770000000000.  
[2023-03-27 20:41:11] local.DEBUG: No budget limit exist. Must create one for auto-budget #4  
[2023-03-27 20:41:11] local.DEBUG: Amount is overruled and will be set to 239.770000000000  
[2023-03-27 20:41:11] local.DEBUG: Trigger budget limit event.  
[2023-03-27 20:41:11] local.DEBUG: Because budget limit #87 had its amount changed to 239.770000000000, available budget limit #14 will be updated.  
[2023-03-27 20:41:11] local.DEBUG: Created budget limit #87.  
[2023-03-27 20:41:11] local.DEBUG: Done with auto budget #4  
[2023-03-27 20:41:11] local.DEBUG: Done with auto budget #4  
```

**Comment by ashemsay** [comment]
Thank you for your quick answer.
So I enabled the debug logs and ran the cron job, I get this:
```
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Now in FireflyIII\Api\V1\Controllers\System\CronController::cron
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Date is 2023-03-26T22:00:00.000000Z
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Now in FireflyIII\Support\Cronjobs\RecurringCronjob::fire
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: It has been 1 s avant since the recurring transactions cron-job has fired. It will fire now!
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Will now fire recurring cron job task for date "2023-03-27 00:00:00".
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Created new CreateRecurringTransactions("2023-03-27")
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Now at start of CreateRecurringTransactions() job for Mon 27 Mar 2023.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Grab all recurrences from the database.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Count of collection is 0
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Left after filtering is 0
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Now running report thing.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Done with handle()
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Marked the last time this job has run as "2023-03-27 00:00:00" (1679868000)
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Done with recurring cron job task.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: It has been 1 s avant since the auto budget cron-job has fired. It will fire now!
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Will now fire auto budget cron job task for date "2023-03-27".
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Created new CreateAutoBudgetLimits("2023-03-27")
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Now at start of CreateAutoBudgetLimits() job for Mon 27 Mar 2023.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Found 2 auto budgets.
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Today (2023-03-27) is not a magic day for monthly auto-budget #1 (part of budget #1 "Budget1")
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Done with auto budget #1
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Today (2023-03-27) is not a magic day for monthly auto-budget #2 (part of budget #2 "Budget2")
firefly-commun-app-1  | [2023-03-27 22:27:27] local.DEBUG: Done with auto budget #2
firefly-commun-app-1  | [2023-03-27 22:27:27] local.INFO: Done with auto budget cron job task.
```
Is it an issue that I had not run the cron before today? What is the magic day? Is it the first day of the month?

**Comment by JC5** [comment, marked as answer]
Yes, the first day of the period (month, week, etc) is when rollovers are set. And if you set them yourself before the cronjob runs, they won't be set.

**Reply by Zyadon7** [reply]
Can you explain "And if you set them yourself before the cronjob runs, they won't be set."  Does this mean that if the automation adds 100 to a budget, and I change it to 150, then the next period auto-budget cron will ignore that budget and do nothing?

**Reply by JC5** [reply]
It means that if today, you set a manual budget for sept 1, sept 30, the cron job will not overwrite it.

**Comment by ashemsay** [comment]
I think seeing the budget is not set (since I had not set up the cron job), I must have set it manually, I guess I'll see everything work perfectly next month.
Obviously I should have read the budget section of the documentation in the first place, I went there only after seing an issue and clearly did not understand how it worked.
Thank you for the clarifications on the matter.
